libraries/chain/block_state.cpp

```diff
@@ -60,45 +60,56 @@ block_state_ptr block_state::create_if_genesis_block(const block_state_legacy& b
+   result.header_exts = bsp.header_exts;
+
+   // set block_state data ----
+   result.block = bsp.block;
+   result.strong_digest = result.compute_finality_digest(); // all block_header_state data populated in result at this point
+   result.weak_digest = create_weak_digest(result.strong_digest);
+
+   // TODO: https://github.com/AntelopeIO/leap/issues/2057
+   // TODO: Do not aggregate votes on blocks created from block_state_legacy. This can be removed when #2057 complete.
+   result.pending_qc = pending_quorum_certificate{result.active_finalizer_policy->finalizers.size(), result.active_finalizer_policy->threshold, result.active_finalizer_policy->max_weak_sum_before_weak_final()};
+   result.valid_qc = {}; // best qc received from the network inside block extension, empty until first savanna proper IF block
+
    // Calculate Merkle tree root in Savanna way so that it is stored in Leaf Node when building block_state.
    auto digests = *bsp.action_receipt_digests_savanna;
    auto action_mroot_svnn = calculate_merkle(std::move(digests));
+
    // built leaf_node and validation_tree
    valid_t::finality_leaf_node_t leaf_node {
       .block_num       = bsp.block_num(),
-      .finality_digest = digest_type{},
+      .finality_digest = result.strong_digest,
       .action_mroot    = action_mroot_svnn
    };
+   // construct valid structure
    incremental_merkle_tree validation_tree;
    validation_tree.append(fc::sha256::hash(leaf_node));
-
-   // construct valid structure
    result.valid = valid_t {
       .validation_tree   = validation_tree,
       .validation_mroots = { validation_tree.get_root() }
    };
 
-   assert(result.block->contains_header_extension(instant_finality_extension::extension_id())); // required by transition mechanism
-   instant_finality_extension if_ext = result.block->extract_header_extension<instant_finality_extension>();
-   assert(if_ext.new_finalizer_policy); // required by transition mechanism
-   result.active_finalizer_policy = std::make_shared<finalizer_policy>(*if_ext.new_finalizer_policy);
-   result.active_proposer_policy = std::make_shared<proposer_policy>();
-   result.active_proposer_policy->active_time = bsp.timestamp();
-   result.active_proposer_policy->proposer_schedule = bsp.active_schedule;
-   // TODO: https://github.com/AntelopeIO/leap/issues/2057
-   // TODO: Do not aggregate votes on blocks created from block_state_legacy. This can be removed when #2057 complete.
-   result.pending_qc = pending_quorum_certificate{result.active_finalizer_policy->finalizers.size(), result.active_finalizer_policy->threshold, result.active_finalizer_policy->max_weak_sum_before_weak_final()};
    result.validated = bsp.is_valid();
    result.pub_keys_recovered = bsp._pub_keys_recovered;
    result.cached_trxs = bsp._cached_trxs;
    result.action_mroot = action_mroot_svnn;
+   result.base_digest = {}; // calculated on demand in get_finality_data()
```

Comment:
is it useful?

Reply:
Yes, Areg wanted all attributes accounted for explicitly.